sidecar/src/backup/pv.rs

```diff
@@ -0,0 +1,92 @@
+use std::fs::{metadata, read_dir, remove_file};
+use std::ops::{AddAssign, Div};
+use std::path::{Path, PathBuf};
+use std::time::{Duration, SystemTime};
+
+use anyhow::{anyhow, Result};
+use async_trait::async_trait;
+use futures::StreamExt;
+use tokio::fs::File;
+use tokio::io;
+use tokio::io::AsyncWriteExt;
+use tonic::Streaming;
+use tracing::debug;
+use xlineapi::SnapshotResponse;
+
+use crate::backup::{Metadata, Provider, SNAPSHOT_SUFFIX};
+
+/// The persistent volume backup
+#[derive(Debug)]
+pub(crate) struct Pv {
+    /// The backup path
+    pub(crate) backup_path: PathBuf,
+}
+
+#[async_trait]
+impl Provider for Pv {
+    async fn latest(&self) -> Result<Option<Metadata>> {
+        let entries = read_dir(&self.backup_path)?;
+        let snapshot = entries
+            .filter_map(|item| {
+                item.ok()
+                    .and_then(|entry| entry.file_name().into_string().ok())
+            })
+            .filter(|item| item.ends_with(SNAPSHOT_SUFFIX))
+            .map(|item| Metadata::try_from(Path::new(&item)))
+            .collect::<Result<Vec<_>, _>>()?
+            .into_iter()
+            .max_by_key(|metadata| metadata.revision);
+        Ok(snapshot)
+    }
+
+    async fn save(&self, mut src: Streaming<SnapshotResponse>, metadata: &Metadata) -> Result<()> {
+        let filename = metadata.to_string();
+        let mut dst = File::create(Path::new(&self.backup_path).join(&filename)).await?;
+        let Some(item) = src.next().await else {
+            return Err(anyhow!("get the item from source failed"))
+        };
+        let mut item = item?;
+        let mut size = item.blob.len();
+        dst.write_all(&item.blob).await?;
+        while item.remaining_bytes > 0 {
+            let Some(got) = src.next().await else {
+                return Err(anyhow!("get the item from source failed"))
+            };
+            item = got?;
+            size.add_assign(item.blob.len());
+            dst.write_all(&item.blob).await?;
```

Comment:
I think this should be append to file. And we can write a snapshot to memory and flush it to disk at one time for better performance.

Reply:
done